<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="generator" content="pdoc 11.0.0"/>
    <title>src API documentation</title>

    <style>/*! * Bootstrap Reboot v5.0.0 (https://getbootstrap.com/) * Copyright 2011-2021 The Bootstrap Authors * Copyright 2011-2021 Twitter, Inc. * Licensed under MIT (https://github.com/twbs/bootstrap/blob/main/LICENSE) * Forked from Normalize.css, licensed MIT (https://github.com/necolas/normalize.css/blob/master/LICENSE.md) */*,::after,::before{box-sizing:border-box}@media (prefers-reduced-motion:no-preference){:root{scroll-behavior:smooth}}body{margin:0;font-family:system-ui,-apple-system,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans","Liberation Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";font-size:1rem;font-weight:400;line-height:1.5;color:#212529;background-color:#fff;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:transparent}hr{margin:1rem 0;color:inherit;background-color:currentColor;border:0;opacity:.25}hr:not([size]){height:1px}h1,h2,h3,h4,h5,h6{margin-top:0;margin-bottom:.5rem;font-weight:500;line-height:1.2}h1{font-size:calc(1.375rem + 1.5vw)}@media (min-width:1200px){h1{font-size:2.5rem}}h2{font-size:calc(1.325rem + .9vw)}@media (min-width:1200px){h2{font-size:2rem}}h3{font-size:calc(1.3rem + .6vw)}@media (min-width:1200px){h3{font-size:1.75rem}}h4{font-size:calc(1.275rem + .3vw)}@media (min-width:1200px){h4{font-size:1.5rem}}h5{font-size:1.25rem}h6{font-size:1rem}p{margin-top:0;margin-bottom:1rem}abbr[data-bs-original-title],abbr[title]{-webkit-text-decoration:underline dotted;text-decoration:underline dotted;cursor:help;-webkit-text-decoration-skip-ink:none;text-decoration-skip-ink:none}address{margin-bottom:1rem;font-style:normal;line-height:inherit}ol,ul{padding-left:2rem}dl,ol,ul{margin-top:0;margin-bottom:1rem}ol ol,ol ul,ul ol,ul ul{margin-bottom:0}dt{font-weight:700}dd{margin-bottom:.5rem;margin-left:0}blockquote{margin:0 0 1rem}b,strong{font-weight:bolder}small{font-size:.875em}mark{padding:.2em;background-color:#fcf8e3}sub,sup{position:relative;font-size:.75em;line-height:0;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}a{color:#0d6efd;text-decoration:underline}a:hover{color:#0a58ca}a:not([href]):not([class]),a:not([href]):not([class]):hover{color:inherit;text-decoration:none}code,kbd,pre,samp{font-family:SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;font-size:1em;direction:ltr;unicode-bidi:bidi-override}pre{display:block;margin-top:0;margin-bottom:1rem;overflow:auto;font-size:.875em}pre code{font-size:inherit;color:inherit;word-break:normal}code{font-size:.875em;color:#d63384;word-wrap:break-word}a>code{color:inherit}kbd{padding:.2rem .4rem;font-size:.875em;color:#fff;background-color:#212529;border-radius:.2rem}kbd kbd{padding:0;font-size:1em;font-weight:700}figure{margin:0 0 1rem}img,svg{vertical-align:middle}table{caption-side:bottom;border-collapse:collapse}caption{padding-top:.5rem;padding-bottom:.5rem;color:#6c757d;text-align:left}th{text-align:inherit;text-align:-webkit-match-parent}tbody,td,tfoot,th,thead,tr{border-color:inherit;border-style:solid;border-width:0}label{display:inline-block}button{border-radius:0}button:focus:not(:focus-visible){outline:0}button,input,optgroup,select,textarea{margin:0;font-family:inherit;font-size:inherit;line-height:inherit}button,select{text-transform:none}[role=button]{cursor:pointer}select{word-wrap:normal}select:disabled{opacity:1}[list]::-webkit-calendar-picker-indicator{display:none}[type=button],[type=reset],[type=submit],button{-webkit-appearance:button}[type=button]:not(:disabled),[type=reset]:not(:disabled),[type=submit]:not(:disabled),button:not(:disabled){cursor:pointer}::-moz-focus-inner{padding:0;border-style:none}textarea{resize:vertical}fieldset{min-width:0;padding:0;margin:0;border:0}legend{float:left;width:100%;padding:0;margin-bottom:.5rem;font-size:calc(1.275rem + .3vw);line-height:inherit}@media (min-width:1200px){legend{font-size:1.5rem}}legend+*{clear:left}::-webkit-datetime-edit-day-field,::-webkit-datetime-edit-fields-wrapper,::-webkit-datetime-edit-hour-field,::-webkit-datetime-edit-minute,::-webkit-datetime-edit-month-field,::-webkit-datetime-edit-text,::-webkit-datetime-edit-year-field{padding:0}::-webkit-inner-spin-button{height:auto}[type=search]{outline-offset:-2px;-webkit-appearance:textfield}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-color-swatch-wrapper{padding:0}::file-selector-button{font:inherit}::-webkit-file-upload-button{font:inherit;-webkit-appearance:button}output{display:inline-block}iframe{border:0}summary{display:list-item;cursor:pointer}progress{vertical-align:baseline}[hidden]{display:none!important}</style>
    <style>/*! syntax-highlighting.css */pre{line-height:125%;}span.linenos{color:inherit; background-color:transparent; padding-left:5px; padding-right:20px;}.pdoc-code .hll{background-color:#ffffcc}.pdoc-code{background:#f8f8f8;}.pdoc-code .c{color:#3D7B7B; font-style:italic}.pdoc-code .err{border:1px solid #FF0000}.pdoc-code .k{color:#008000; font-weight:bold}.pdoc-code .o{color:#666666}.pdoc-code .ch{color:#3D7B7B; font-style:italic}.pdoc-code .cm{color:#3D7B7B; font-style:italic}.pdoc-code .cp{color:#9C6500}.pdoc-code .cpf{color:#3D7B7B; font-style:italic}.pdoc-code .c1{color:#3D7B7B; font-style:italic}.pdoc-code .cs{color:#3D7B7B; font-style:italic}.pdoc-code .gd{color:#A00000}.pdoc-code .ge{font-style:italic}.pdoc-code .gr{color:#E40000}.pdoc-code .gh{color:#000080; font-weight:bold}.pdoc-code .gi{color:#008400}.pdoc-code .go{color:#717171}.pdoc-code .gp{color:#000080; font-weight:bold}.pdoc-code .gs{font-weight:bold}.pdoc-code .gu{color:#800080; font-weight:bold}.pdoc-code .gt{color:#0044DD}.pdoc-code .kc{color:#008000; font-weight:bold}.pdoc-code .kd{color:#008000; font-weight:bold}.pdoc-code .kn{color:#008000; font-weight:bold}.pdoc-code .kp{color:#008000}.pdoc-code .kr{color:#008000; font-weight:bold}.pdoc-code .kt{color:#B00040}.pdoc-code .m{color:#666666}.pdoc-code .s{color:#BA2121}.pdoc-code .na{color:#687822}.pdoc-code .nb{color:#008000}.pdoc-code .nc{color:#0000FF; font-weight:bold}.pdoc-code .no{color:#880000}.pdoc-code .nd{color:#AA22FF}.pdoc-code .ni{color:#717171; font-weight:bold}.pdoc-code .ne{color:#CB3F38; font-weight:bold}.pdoc-code .nf{color:#0000FF}.pdoc-code .nl{color:#767600}.pdoc-code .nn{color:#0000FF; font-weight:bold}.pdoc-code .nt{color:#008000; font-weight:bold}.pdoc-code .nv{color:#19177C}.pdoc-code .ow{color:#AA22FF; font-weight:bold}.pdoc-code .w{color:#bbbbbb}.pdoc-code .mb{color:#666666}.pdoc-code .mf{color:#666666}.pdoc-code .mh{color:#666666}.pdoc-code .mi{color:#666666}.pdoc-code .mo{color:#666666}.pdoc-code .sa{color:#BA2121}.pdoc-code .sb{color:#BA2121}.pdoc-code .sc{color:#BA2121}.pdoc-code .dl{color:#BA2121}.pdoc-code .sd{color:#BA2121; font-style:italic}.pdoc-code .s2{color:#BA2121}.pdoc-code .se{color:#AA5D1F; font-weight:bold}.pdoc-code .sh{color:#BA2121}.pdoc-code .si{color:#A45A77; font-weight:bold}.pdoc-code .sx{color:#008000}.pdoc-code .sr{color:#A45A77}.pdoc-code .s1{color:#BA2121}.pdoc-code .ss{color:#19177C}.pdoc-code .bp{color:#008000}.pdoc-code .fm{color:#0000FF}.pdoc-code .vc{color:#19177C}.pdoc-code .vg{color:#19177C}.pdoc-code .vi{color:#19177C}.pdoc-code .vm{color:#19177C}.pdoc-code .il{color:#666666}</style>
    <style>/*! theme.css */:root{--pdoc-background:#fff;}.pdoc{--text:#212529;--muted:#6c757d;--link:#3660a5;--link-hover:#1659c5;--code:#f8f8f8;--active:#fff598;--accent:#eee;--accent2:#c1c1c1;--nav-hover:rgba(255, 255, 255, 0.5);--name:#0066BB;--def:#008800;--annotation:#007020;}</style>
    <style>/*! layout.css */html, body{width:100%;height:100%;}html, main{scroll-behavior:smooth;}body{background-color:var(--pdoc-background);}@media (max-width:769px){#navtoggle{cursor:pointer;position:absolute;width:50px;height:40px;top:1rem;right:1rem;border-color:var(--text);color:var(--text);display:flex;opacity:0.8;}#navtoggle:hover{opacity:1;}#togglestate + div{display:none;}#togglestate:checked + div{display:inherit;}main, header{padding:2rem 3vw;}header + main{margin-top:-3rem;}.git-button{display:none !important;}nav input[type="search"]{max-width:77%;}nav input[type="search"]:first-child{margin-top:-6px;}nav input[type="search"]:valid ~ *{display:none !important;}}@media (min-width:770px){:root{--sidebar-width:clamp(12.5rem, 28vw, 22rem);}nav{position:fixed;overflow:auto;height:100vh;width:var(--sidebar-width);}main, header{padding:3rem 2rem 3rem calc(var(--sidebar-width) + 3rem);width:calc(54rem + var(--sidebar-width));max-width:100%;}header + main{margin-top:-4rem;}#navtoggle{display:none;}}#togglestate{position:absolute;height:0;}nav.pdoc{--pad:clamp(0.5rem, 2vw, 1.75rem);--indent:1.5rem;background-color:var(--accent);border-right:1px solid var(--accent2);box-shadow:0 0 20px rgba(50, 50, 50, .2) inset;padding:0 0 0 var(--pad);overflow-wrap:anywhere;scrollbar-width:thin; scrollbar-color:var(--accent2) transparent }nav.pdoc::-webkit-scrollbar{width:.4rem; }nav.pdoc::-webkit-scrollbar-thumb{background-color:var(--accent2); }nav.pdoc > div{padding:var(--pad) 0;}nav.pdoc .module-list-button{display:inline-flex;align-items:center;color:var(--text);border-color:var(--muted);margin-bottom:1rem;}nav.pdoc .module-list-button:hover{border-color:var(--text);}nav.pdoc input[type=search]{display:block;outline-offset:0;width:calc(100% - var(--pad));}nav.pdoc .logo{max-width:calc(100% - var(--pad));max-height:35vh;display:block;margin:0 auto 1rem;transform:translate(calc(-.5 * var(--pad)), 0);}nav.pdoc ul{list-style:none;padding-left:0;}nav.pdoc > div > ul{margin-left:calc(0px - var(--pad));}nav.pdoc li a{padding:.2rem 0 .2rem calc(var(--pad) + var(--indent));}nav.pdoc > div > ul > li > a{padding-left:var(--pad);}nav.pdoc li{transition:all 100ms;}nav.pdoc li:hover{background-color:var(--nav-hover);}nav.pdoc a, nav.pdoc a:hover{color:var(--text);}nav.pdoc a{display:block;}nav.pdoc > h2:first-of-type{margin-top:1.5rem;}nav.pdoc .class:before{content:"class ";color:var(--muted);}nav.pdoc .function:after{content:"()";color:var(--muted);}nav.pdoc footer:before{content:"";display:block;width:calc(100% - var(--pad));border-top:solid var(--accent2) 1px;margin-top:1.5rem;padding-top:.5rem;}nav.pdoc footer{font-size:small;}</style>
    <style>/*! content.css */.pdoc{color:var(--text);box-sizing:border-box;line-height:1.5;background:none;}.pdoc .pdoc-button{display:inline-block;border:solid black 1px;border-radius:2px;font-size:.75rem;padding:calc(0.5em - 1px) 1em;transition:100ms all;}.pdoc .pdoc-alert{padding:1rem 1rem 1rem calc(1.5rem + 24px);border:1px solid transparent;border-radius:.25rem;background-repeat:no-repeat;background-position:1rem center;margin-bottom:1rem;}.pdoc .pdoc-alert > *:last-child{margin-bottom:0;}.pdoc .pdoc-alert-note {color:#084298;background-color:#cfe2ff;border-color:#b6d4fe;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23084298%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M8%2016A8%208%200%201%200%208%200a8%208%200%200%200%200%2016zm.93-9.412-1%204.705c-.07.34.029.533.304.533.194%200%20.487-.07.686-.246l-.088.416c-.287.346-.92.598-1.465.598-.703%200-1.002-.422-.808-1.319l.738-3.468c.064-.293.006-.399-.287-.47l-.451-.081.082-.381%202.29-.287zM8%205.5a1%201%200%201%201%200-2%201%201%200%200%201%200%202z%22/%3E%3C/svg%3E");}.pdoc .pdoc-alert-warning{color:#664d03;background-color:#fff3cd;border-color:#ffecb5;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23664d03%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M8.982%201.566a1.13%201.13%200%200%200-1.96%200L.165%2013.233c-.457.778.091%201.767.98%201.767h13.713c.889%200%201.438-.99.98-1.767L8.982%201.566zM8%205c.535%200%20.954.462.9.995l-.35%203.507a.552.552%200%200%201-1.1%200L7.1%205.995A.905.905%200%200%201%208%205zm.002%206a1%201%200%201%201%200%202%201%201%200%200%201%200-2z%22/%3E%3C/svg%3E");}.pdoc .pdoc-alert-danger{color:#842029;background-color:#f8d7da;border-color:#f5c2c7;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23842029%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M5.52.359A.5.5%200%200%201%206%200h4a.5.5%200%200%201%20.474.658L8.694%206H12.5a.5.5%200%200%201%20.395.807l-7%209a.5.5%200%200%201-.873-.454L6.823%209.5H3.5a.5.5%200%200%201-.48-.641l2.5-8.5z%22/%3E%3C/svg%3E");}.pdoc .visually-hidden{position:absolute !important;width:1px !important;height:1px !important;padding:0 !important;margin:-1px !important;overflow:hidden !important;clip:rect(0, 0, 0, 0) !important;white-space:nowrap !important;border:0 !important;}.pdoc h1, .pdoc h2, .pdoc h3{font-weight:300;margin:.3em 0;padding:.2em 0;}.pdoc a{text-decoration:none;color:var(--link);}.pdoc a:hover{color:var(--link-hover);}.pdoc blockquote{margin-left:2rem;}.pdoc pre{border-top:1px solid var(--accent2);border-bottom:1px solid var(--accent2);margin-top:0;margin-bottom:1em;padding:.5rem 0 .5rem .5rem;overflow-x:auto;background-color:var(--code);}.pdoc code{color:var(--text);padding:.2em .4em;margin:0;font-size:85%;background-color:var(--code);border-radius:6px;}.pdoc a > code{color:inherit;}.pdoc pre > code{display:inline-block;font-size:inherit;background:none;border:none;padding:0;}.pdoc > section:not(:first-of-type){margin-bottom:1.5rem;}.pdoc .modulename{margin-top:0;font-weight:bold;}.pdoc .modulename a{color:var(--link);transition:100ms all;}.pdoc .git-button{float:right;border:solid var(--link) 1px;}.pdoc .git-button:hover{background-color:var(--link);color:var(--pdoc-background);}.pdoc details{filter:opacity(1);}.pdoc details:not([open]){height:0;}.pdoc details > summary{position:absolute;top:-35px;right:0;font-size:.75rem;color:var(--muted);padding:0 .7em;user-select:none;cursor:pointer;}.pdoc details > summary:focus{outline:0;}.pdoc > section:first-of-type details > summary{top:-20px;}.pdoc .docstring{margin-bottom:1.5rem;}.pdoc .docstring .pdoc-code{margin-left:1em;margin-right:1em;}.pdoc h1:target,.pdoc h2:target,.pdoc h3:target,.pdoc h4:target,.pdoc h5:target,.pdoc h6:target{background-color:var(--active);box-shadow:-1rem 0 0 0 var(--active);}.pdoc div:target > .attr,.pdoc section:target > .attr,.pdoc dd:target > a{background-color:var(--active);}.pdoc .attr:hover{filter:contrast(0.95);}.pdoc .headerlink{position:absolute;width:0;margin-left:-1.5rem;line-height:1.4rem;font-size:1.5rem;font-weight:normal;transition:all 100ms ease-in-out;opacity:0;user-select:none;}.pdoc .attr > .headerlink{margin-left:-2.5rem;}.pdoc *:hover > .headerlink,.pdoc *:target > .attr > .headerlink{opacity:1;}.pdoc .attr{display:block;color:var(--text);margin:.5rem 0 .5rem;padding:.4rem 5rem .4rem 1rem;background-color:var(--accent);}.pdoc .classattr{margin-left:2rem;}.pdoc .name{color:var(--name);font-weight:bold;}.pdoc .def{color:var(--def);font-weight:bold;}.pdoc .signature{white-space:pre-wrap;}.pdoc .annotation{color:var(--annotation);}.pdoc .inherited{margin-left:2rem;}.pdoc .inherited dt{font-weight:700;}.pdoc .inherited dt, .pdoc .inherited dd{display:inline;margin-left:0;margin-bottom:.5rem;}.pdoc .inherited dd:not(:last-child):after{content:", ";}.pdoc .inherited .class:before{content:"class ";}.pdoc .inherited .function a:after{content:"()";}.pdoc .search-result .docstring{overflow:auto;max-height:25vh;}.pdoc .search-result.focused > .attr{background-color:var(--active);}.pdoc .attribution{margin-top:2rem;display:block;opacity:0.5;transition:all 200ms;filter:grayscale(100%);}.pdoc .attribution:hover{opacity:1;filter:grayscale(0%);}.pdoc .attribution img{margin-left:5px;height:35px;vertical-align:middle;width:70px;transition:all 200ms;}.pdoc table{display:block;width:max-content;max-width:100%;overflow:auto;margin-bottom:1rem;}.pdoc table th{font-weight:600;}.pdoc table th, .pdoc table td{padding:6px 13px;border:1px solid var(--accent2);}</style>
    <style>/*! custom.css */</style></head>
<body>
    <nav class="pdoc">
        <label id="navtoggle" for="togglestate" class="pdoc-button"><svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 30 30'><path stroke-linecap='round' stroke="currentColor" stroke-miterlimit='10' stroke-width='2' d='M4 7h22M4 15h22M4 23h22'/></svg></label>
        <input id="togglestate" type="checkbox" aria-hidden="true" tabindex="-1">
        <div>




        <h2>API Documentation</h2>
            <ul class="memberlist">
            <li>
                    <a class="class" href="#RunConfig">RunConfig</a>
                            <ul class="memberlist">
                        <li>
                                <a class="function" href="#RunConfig.__init__">RunConfig</a>
                        </li>
                        <li>
                                <a class="function" href="#RunConfig.verify_arguments">verify_arguments</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="function" href="#map_reads">map_reads</a>
            </li>
            <li>
                    <a class="function" href="#sort_bam">sort_bam</a>
            </li>
            <li>
                    <a class="function" href="#add_rg">add_rg</a>
            </li>
            <li>
                    <a class="function" href="#make_rg_bamlist">make_rg_bamlist</a>
            </li>
            <li>
                    <a class="function" href="#merge_bams">merge_bams</a>
            </li>
            <li>
                    <a class="function" href="#rename_to_merged_bam">rename_to_merged_bam</a>
            </li>
            <li>
                    <a class="function" href="#mark_dups">mark_dups</a>
            </li>
            <li>
                    <a class="function" href="#index_bam">index_bam</a>
            </li>
            <li>
                    <a class="function" href="#realigner_target_creator">realigner_target_creator</a>
            </li>
            <li>
                    <a class="function" href="#indel_realigner">indel_realigner</a>
            </li>
            <li>
                    <a class="function" href="#print_sample_pipeline">print_sample_pipeline</a>
            </li>
    </ul>



        <a class="attribution" title="pdoc: Python API documentation generator" href="https://pdoc.dev">
            built with <span class="visually-hidden">pdoc</span><img
                alt="pdoc logo"
                src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20role%3D%22img%22%20aria-label%3D%22pdoc%20logo%22%20width%3D%22300%22%20height%3D%22150%22%20viewBox%3D%22-1%200%2060%2030%22%3E%3Ctitle%3Epdoc%3C/title%3E%3Cpath%20d%3D%22M29.621%2021.293c-.011-.273-.214-.475-.511-.481a.5.5%200%200%200-.489.503l-.044%201.393c-.097.551-.695%201.215-1.566%201.704-.577.428-1.306.486-2.193.182-1.426-.617-2.467-1.654-3.304-2.487l-.173-.172a3.43%203.43%200%200%200-.365-.306.49.49%200%200%200-.286-.196c-1.718-1.06-4.931-1.47-7.353.191l-.219.15c-1.707%201.187-3.413%202.131-4.328%201.03-.02-.027-.49-.685-.141-1.763.233-.721.546-2.408.772-4.076.042-.09.067-.187.046-.288.166-1.347.277-2.625.241-3.351%201.378-1.008%202.271-2.586%202.271-4.362%200-.976-.272-1.935-.788-2.774-.057-.094-.122-.18-.184-.268.033-.167.052-.339.052-.516%200-1.477-1.202-2.679-2.679-2.679-.791%200-1.496.352-1.987.9a6.3%206.3%200%200%200-1.001.029c-.492-.564-1.207-.929-2.012-.929-1.477%200-2.679%201.202-2.679%202.679A2.65%202.65%200%200%200%20.97%206.554c-.383.747-.595%201.572-.595%202.41%200%202.311%201.507%204.29%203.635%205.107-.037.699-.147%202.27-.423%203.294l-.137.461c-.622%202.042-2.515%208.257%201.727%2010.643%201.614.908%203.06%201.248%204.317%201.248%202.665%200%204.492-1.524%205.322-2.401%201.476-1.559%202.886-1.854%206.491.82%201.877%201.393%203.514%201.753%204.861%201.068%202.223-1.713%202.811-3.867%203.399-6.374.077-.846.056-1.469.054-1.537zm-4.835%204.313c-.054.305-.156.586-.242.629-.034-.007-.131-.022-.307-.157-.145-.111-.314-.478-.456-.908.221.121.432.25.675.355.115.039.219.051.33.081zm-2.251-1.238c-.05.33-.158.648-.252.694-.022.001-.125-.018-.307-.157-.217-.166-.488-.906-.639-1.573.358.344.754.693%201.198%201.036zm-3.887-2.337c-.006-.116-.018-.231-.041-.342.635.145%201.189.368%201.599.625.097.231.166.481.174.642-.03.049-.055.101-.067.158-.046.013-.128.026-.298.004-.278-.037-.901-.57-1.367-1.087zm-1.127-.497c.116.306.176.625.12.71-.019.014-.117.045-.345.016-.206-.027-.604-.332-.986-.695.41-.051.816-.056%201.211-.031zm-4.535%201.535c.209.22.379.47.358.598-.006.041-.088.138-.351.234-.144.055-.539-.063-.979-.259a11.66%2011.66%200%200%200%20.972-.573zm.983-.664c.359-.237.738-.418%201.126-.554.25.237.479.548.457.694-.006.042-.087.138-.351.235-.174.064-.694-.105-1.232-.375zm-3.381%201.794c-.022.145-.061.29-.149.401-.133.166-.358.248-.69.251h-.002c-.133%200-.306-.26-.45-.621.417.091.854.07%201.291-.031zm-2.066-8.077a4.78%204.78%200%200%201-.775-.584c.172-.115.505-.254.88-.378l-.105.962zm-.331%202.302a10.32%2010.32%200%200%201-.828-.502c.202-.143.576-.328.984-.49l-.156.992zm-.45%202.157l-.701-.403c.214-.115.536-.249.891-.376a11.57%2011.57%200%200%201-.19.779zm-.181%201.716c.064.398.194.702.298.893-.194-.051-.435-.162-.736-.398.061-.119.224-.3.438-.495zM8.87%204.141c0%20.152-.123.276-.276.276s-.275-.124-.275-.276.123-.276.276-.276.275.124.275.276zm-.735-.389a1.15%201.15%200%200%200-.314.783%201.16%201.16%200%200%200%201.162%201.162c.457%200%20.842-.27%201.032-.653.026.117.042.238.042.362a1.68%201.68%200%200%201-1.679%201.679%201.68%201.68%200%200%201-1.679-1.679c0-.843.626-1.535%201.436-1.654zM5.059%205.406A1.68%201.68%200%200%201%203.38%207.085a1.68%201.68%200%200%201-1.679-1.679c0-.037.009-.072.011-.109.21.3.541.508.935.508a1.16%201.16%200%200%200%201.162-1.162%201.14%201.14%200%200%200-.474-.912c.015%200%20.03-.005.045-.005.926.001%201.679.754%201.679%201.68zM3.198%204.141c0%20.152-.123.276-.276.276s-.275-.124-.275-.276.123-.276.276-.276.275.124.275.276zM1.375%208.964c0-.52.103-1.035.288-1.52.466.394%201.06.64%201.717.64%201.144%200%202.116-.725%202.499-1.738.383%201.012%201.355%201.738%202.499%201.738.867%200%201.631-.421%202.121-1.062.307.605.478%201.267.478%201.942%200%202.486-2.153%204.51-4.801%204.51s-4.801-2.023-4.801-4.51zm24.342%2019.349c-.985.498-2.267.168-3.813-.979-3.073-2.281-5.453-3.199-7.813-.705-1.315%201.391-4.163%203.365-8.423.97-3.174-1.786-2.239-6.266-1.261-9.479l.146-.492c.276-1.02.395-2.457.444-3.268a6.11%206.11%200%200%200%201.18.115%206.01%206.01%200%200%200%202.536-.562l-.006.175c-.802.215-1.848.612-2.021%201.25-.079.295.021.601.274.837.219.203.415.364.598.501-.667.304-1.243.698-1.311%201.179-.02.144-.022.507.393.787.213.144.395.26.564.365-1.285.521-1.361.96-1.381%201.126-.018.142-.011.496.427.746l.854.489c-.473.389-.971.914-.999%201.429-.018.278.095.532.316.713.675.556%201.231.721%201.653.721.059%200%20.104-.014.158-.02.207.707.641%201.64%201.513%201.64h.013c.8-.008%201.236-.345%201.462-.626.173-.216.268-.457.325-.692.424.195.93.374%201.372.374.151%200%20.294-.021.423-.068.732-.27.944-.704.993-1.021.009-.061.003-.119.002-.179.266.086.538.147.789.147.15%200%20.294-.021.423-.069.542-.2.797-.489.914-.754.237.147.478.258.704.288.106.014.205.021.296.021.356%200%20.595-.101.767-.229.438.435%201.094.992%201.656%201.067.106.014.205.021.296.021a1.56%201.56%200%200%200%20.323-.035c.17.575.453%201.289.866%201.605.358.273.665.362.914.362a.99.99%200%200%200%20.421-.093%201.03%201.03%200%200%200%20.245-.164c.168.428.39.846.68%201.068.358.273.665.362.913.362a.99.99%200%200%200%20.421-.093c.317-.148.512-.448.639-.762.251.157.495.257.726.257.127%200%20.25-.024.37-.071.427-.17.706-.617.841-1.314.022-.015.047-.022.068-.038.067-.051.133-.104.196-.159-.443%201.486-1.107%202.761-2.086%203.257zM8.66%209.925a.5.5%200%201%200-1%200c0%20.653-.818%201.205-1.787%201.205s-1.787-.552-1.787-1.205a.5.5%200%201%200-1%200c0%201.216%201.25%202.205%202.787%202.205s2.787-.989%202.787-2.205zm4.4%2015.965l-.208.097c-2.661%201.258-4.708%201.436-6.086.527-1.542-1.017-1.88-3.19-1.844-4.198a.4.4%200%200%200-.385-.414c-.242-.029-.406.164-.414.385-.046%201.249.367%203.686%202.202%204.896.708.467%201.547.7%202.51.7%201.248%200%202.706-.392%204.362-1.174l.185-.086a.4.4%200%200%200%20.205-.527c-.089-.204-.326-.291-.527-.206zM9.547%202.292c.093.077.205.114.317.114a.5.5%200%200%200%20.318-.886L8.817.397a.5.5%200%200%200-.703.068.5.5%200%200%200%20.069.703l1.364%201.124zm-7.661-.065c.086%200%20.173-.022.253-.068l1.523-.893a.5.5%200%200%200-.506-.863l-1.523.892a.5.5%200%200%200-.179.685c.094.158.261.247.432.247z%22%20transform%3D%22matrix%28-1%200%200%201%2058%200%29%22%20fill%3D%22%233bb300%22/%3E%3Cpath%20d%3D%22M.3%2021.86V10.18q0-.46.02-.68.04-.22.18-.5.28-.54%201.34-.54%201.06%200%201.42.28.38.26.44.78.76-1.04%202.38-1.04%201.64%200%203.1%201.54%201.46%201.54%201.46%203.58%200%202.04-1.46%203.58-1.44%201.54-3.08%201.54-1.64%200-2.38-.92v4.04q0%20.46-.04.68-.02.22-.18.5-.14.3-.5.42-.36.12-.98.12-.62%200-1-.12-.36-.12-.52-.4-.14-.28-.18-.5-.02-.22-.02-.68zm3.96-9.42q-.46.54-.46%201.18%200%20.64.46%201.18.48.52%201.2.52.74%200%201.24-.52.52-.52.52-1.18%200-.66-.48-1.18-.48-.54-1.26-.54-.76%200-1.22.54zm14.741-8.36q.16-.3.54-.42.38-.12%201-.12.64%200%201.02.12.38.12.52.42.16.3.18.54.04.22.04.68v11.94q0%20.46-.04.7-.02.22-.18.5-.3.54-1.7.54-1.38%200-1.54-.98-.84.96-2.34.96-1.8%200-3.28-1.56-1.48-1.58-1.48-3.66%200-2.1%201.48-3.68%201.5-1.58%203.28-1.58%201.48%200%202.3%201v-4.2q0-.46.02-.68.04-.24.18-.52zm-3.24%2010.86q.52.54%201.26.54.74%200%201.22-.54.5-.54.5-1.18%200-.66-.48-1.22-.46-.56-1.26-.56-.8%200-1.28.56-.48.54-.48%201.2%200%20.66.52%201.2zm7.833-1.2q0-2.4%201.68-3.96%201.68-1.56%203.84-1.56%202.16%200%203.82%201.56%201.66%201.54%201.66%203.94%200%201.66-.86%202.96-.86%201.28-2.1%201.9-1.22.6-2.54.6-1.32%200-2.56-.64-1.24-.66-2.1-1.92-.84-1.28-.84-2.88zm4.18%201.44q.64.48%201.3.48.66%200%201.32-.5.66-.5.66-1.48%200-.98-.62-1.46-.62-.48-1.34-.48-.72%200-1.34.5-.62.5-.62%201.48%200%20.96.64%201.46zm11.412-1.44q0%20.84.56%201.32.56.46%201.18.46.64%200%201.18-.36.56-.38.9-.38.6%200%201.46%201.06.46.58.46%201.04%200%20.76-1.1%201.42-1.14.8-2.8.8-1.86%200-3.58-1.34-.82-.64-1.34-1.7-.52-1.08-.52-2.36%200-1.3.52-2.34.52-1.06%201.34-1.7%201.66-1.32%203.54-1.32.76%200%201.48.22.72.2%201.06.4l.32.2q.36.24.56.38.52.4.52.92%200%20.5-.42%201.14-.72%201.1-1.38%201.1-.38%200-1.08-.44-.36-.34-1.04-.34-.66%200-1.24.48-.58.48-.58%201.34z%22%20fill%3D%22green%22/%3E%3C/svg%3E"/>
        </a>
</div>
    </nav>
    <main class="pdoc">
            <section>
                    <h1 class="modulename">
src    </h1>

                
                        <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="kn">import</span> <span class="nn">yaml</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">yaml.loader</span> <span class="kn">import</span> <span class="n">FullLoader</span>

<span class="k">class</span> <span class="nc">RunConfig</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class to hold the configuration parameters for a run.</span>
<span class="sd">    </span>
<span class="sd">    Attributes:</span>
<span class="sd">    - sample_dict (dict): A dictionary mapping sample names to dictionaries mapping unique fastq_ids to the file names of the paired-end fastq files (r1 and r2).</span>
<span class="sd">    - samples (list): A list of sample names in the order they appear in the configuration file (within the dictionary sample_dict).</span>
<span class="sd">    - fastq_folder (str): The folder path where the fastq files are located.</span>
<span class="sd">    - reference_fasta (str): The path to the reference fasta file used for alignment.</span>
<span class="sd">    - output_folder (str): The folder path where the output files will be written.</span>
<span class="sd">    - threads (int): The number of threads to use for parallelized tools.</span>
<span class="sd">    - alignment_name (str): A label used to identify the alignment in the output file names.</span>
<span class="sd">    - call_bwa (str): The command or path to the executable for the BWA aligner.</span>
<span class="sd">    - call_samtools (str): The command or path to the executable for the SAMtools toolkit.</span>
<span class="sd">    - call_picard (str): The command or path to the executable for the Picard toolkit.</span>
<span class="sd">    - call_gatk (str): The command or path to the executable for the GATK toolkit.</span>
<span class="sd">    </span>
<span class="sd">    Methods:</span>
<span class="sd">    - __init__(self, config_file): Initializes a new RunConfig object from a YAML configuration file.</span>
<span class="sd">    - verify_arguments(self): Checks if the various elements of the configuration file exist</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">config_file</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes a new RunConfig object from a YAML configuration file.</span>
<span class="sd">        </span>
<span class="sd">        Parameters:</span>
<span class="sd">        - config_file (str): The path to the YAML configuration file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span> <span class="k">as</span> <span class="nb">file</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="n">Loader</span><span class="o">=</span><span class="n">FullLoader</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_dict</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;sample_dict&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;reference_fasta&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_folder</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;output_folder&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">threads</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;threads&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment_name</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;alignment_name&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_bwa</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_bwa&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_samtools</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_samtools&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_picard</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_picard&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_gatk</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_gatk&#39;</span><span class="p">]</span>        

    <span class="k">def</span> <span class="nf">verify_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Checks if the various elements of the configuration file exist</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># check if path to fastqs exists</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">)</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="s2">&quot;reference genome {self.reference_fasta} not found!&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;reference genome found!&quot;</span><span class="p">)</span>

            
<span class="k">def</span> <span class="nf">map_reads</span><span class="p">(</span><span class="n">fastq_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">run</span><span class="p">:</span> <span class="n">RunConfig</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Map reads from a pair of FASTQ files to a reference genome using BWA-MEM and SAMtools.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    - fastq_id (str): The ID of the r1 and r2 FASTQ files pair to map reads from.</span>
<span class="sd">    - sample (str): The name of the sample to map reads for.</span>
<span class="sd">    - run (RunConfig): An object containing the parameters and file paths for the mapping run - created with the RunConfig() class.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list of command-line arguments for running BWA and SAMtools to map reads.</span>
<span class="sd">    </span>
<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; run = RunConfig(config/run.yml)</span>
<span class="sd">    &gt;&gt;&gt; map_reads(&#39;id1&#39;, &#39;sample1&#39;, run)</span>
<span class="sd">    [&#39;/path/to/bwa mem /path/to/reference.fasta /path/to/id1.fastq.gz /path/to/id1_2.fastq.gz -t 8 | /path/to/samtools view -hbS -@ 8 - -o /path/to/sample1_id1.bwa.bam&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># build BWA MEM + SAMTOOLS VIEW command</span>
    <span class="n">bwa_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">run</span><span class="o">.</span><span class="n">call_bwa</span><span class="p">,</span> <span class="s2">&quot;mem&quot;</span><span class="p">,</span> <span class="n">run</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{run.sample_dict[sample][fastq_id][0]}/{run.sample_dict[sample][fastq_id][1]}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{run.sample_dict[sample][fastq_id][0]}/{run.sample_dict[sample][fastq_id][2]}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">),</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;view&quot;</span><span class="p">,</span> <span class="s2">&quot;-hbS&quot;</span><span class="p">,</span> <span class="s2">&quot;-@&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">),</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{run.output_folder}/{sample}_{fastq_id}_{run.alignment_name}.bam&quot;</span>
    <span class="p">]</span>
    
    <span class="k">return</span> <span class="n">bwa_command</span>


<span class="k">def</span> <span class="nf">sort_bam</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_samtools</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sort a BAM file using SAMtools sort command.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): Path to the BAM file to be sorted.</span>
<span class="sd">    - call_samtools (str): Path to the SAMtools executable.</span>
<span class="sd">    - threads (int): Number of threads to use for sorting the BAM file (default=1).</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list of command-line arguments for running the SAMtools sort command to sort the input BAM file.</span>
<span class="sd">    </span>
<span class="sd">    Raises:</span>
<span class="sd">    - TypeError: If the BAM file does not end with &#39;.bam&#39;.</span>

<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; sort_bam(&#39;/path/to/input.bam&#39;, &#39;/path/to/samtools&#39;, 4)</span>
<span class="sd">    [&#39;/path/to/samtools sort -@ 4 /path/to/input.bam -o /path/to/input_sorted.bam&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span> 
    
    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam !&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_sorted.bam&#39;</span>
    
    <span class="c1"># Build SAMTOOLS SORT command</span>
    <span class="n">samtools_sort</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;sort&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-@&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">),</span>
        <span class="n">bam_file</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">out_bam</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">samtools_sort</span>


<span class="k">def</span> <span class="nf">add_rg</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">fastq_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_picard</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add read groups to a BAM file using PicardTools AddOrReplaceReadGroups.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): Path to the input BAM file to add read groups to.</span>
<span class="sd">    - sample (str): Name of the sample to add read groups for.</span>
<span class="sd">    - fastq_id (str): ID of the FASTQ file to add read groups for.</span>
<span class="sd">    - call_picard (str): Path to the PicardTools executable to use.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list of command-line arguments for running PicardTools AddOrReplaceReadGroups.</span>

<span class="sd">    Raises:</span>
<span class="sd">    - TypeError: If the BAM file does not end with &#39;.bam&#39;.</span>
<span class="sd">    </span>
<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; call_picard = &#39;/path/to/picard.jar&#39;</span>
<span class="sd">    &gt;&gt;&gt; bam_file = &#39;/path/to/input.bam&#39;</span>
<span class="sd">    &gt;&gt;&gt; sample = &#39;sample1&#39;</span>
<span class="sd">    &gt;&gt;&gt; fastq_id = &#39;fastq1&#39;</span>
<span class="sd">    &gt;&gt;&gt; add_rg(bam_file, sample, fastq_id, call_picard)</span>
<span class="sd">    [&#39;/path/to/picard.jar&#39;, &#39;AddOrReplaceReadGroups&#39;, &#39;I=/path/to/input.bam&#39;, &#39;O=/path/to/input_rg.bam&#39;, &#39;RGID=fastq1&#39;, &#39;RGLB=sample1_lib&#39;, &#39;RGPL=Illumina&#39;, &#39;RGPU=fastq1&#39;, &#39;RGSM=sample1&#39;, &#39;VALIDATION_STRINGENCY=SILENT&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam !&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_rg.bam&#39;</span>
    
    <span class="c1"># Build PICARDTOOLS AddOrReplaceReadGroups command</span>
    <span class="n">picard_add_rg</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_picard</span><span class="p">,</span> <span class="s2">&quot;AddOrReplaceReadGroups&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;I={bam_file}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;O={out_bam}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGID={fastq_id}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGLB={sample}_lib&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGPL=Illumina&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGPU={fastq_id}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGSM={sample}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;VALIDATION_STRINGENCY=SILENT&quot;</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">picard_add_rg</span>


<span class="k">def</span> <span class="nf">make_rg_bamlist</span><span class="p">(</span><span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Constructs a command to generate a list of BAM files with read group (RG) information</span>
<span class="sd">    for a given sample and output folder.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - sample (str): The name of the sample to generate the BAM file list for.</span>
<span class="sd">    - output_folder (str): The folder where the BAM files are stored.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list containing the ls command to generate the BAM file list.</span>
<span class="sd">    </span>
<span class="sd">    Examples:</span>
<span class="sd">    &gt;&gt;&gt; make_rg_bamlist(&#39;sample1&#39;, &#39;/path/to/output/folder&#39;)</span>
<span class="sd">    [&#39;ls&#39;, &#39;/path/to/output/folder/sample1_*_sorted_rg.bam&#39;, &#39;&gt;&#39;, &#39;/path/to/output/folder/sample1.bam.list&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build ls command</span>
    <span class="n">ls_bamlist</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;ls&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_*_sorted_rg.bam&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}.bam.list&quot;</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">ls_bamlist</span>


<span class="k">def</span> <span class="nf">merge_bams</span><span class="p">(</span><span class="n">bam_list</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_samtools</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">alignment_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Constructs a command to merge multiple BAM files into a single BAM file for a given sample.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - bam_list (str): The path to a file containing a list of BAM files to merge.</span>
<span class="sd">    - sample (str): The name of the sample to merge the BAM files for.</span>
<span class="sd">    - call_samtools (str): The command to call SAMTOOLS.</span>
<span class="sd">    - alignment_name (str): The name of the alignment</span>
<span class="sd">    - threads (int): The number of threads to use for the merging process (default=1).</span>
<span class="sd">    - output_folder (str): The folder where the merged BAM file will be stored.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list containing the SAMTOOLS MERGE command to merge the BAM files.</span>

<span class="sd">    Examples:</span>
<span class="sd">        </span>
<span class="sd">    &gt;&gt;&gt; merge_bams(&quot;/path/to/bam.list&quot;, &quot;sample1&quot;, &quot;samtools&quot;, 8, &quot;/path/to/output/folder&quot;)</span>
<span class="sd">    [&#39;samtools&#39;, &#39;merge&#39;, &#39;-@&#39;, &#39;8&#39;, &#39;-r&#39;, &#39;/path/to/output/folder/sample1_merged.bam&#39;, &#39;-b&#39;, &#39;/path/to/bam.list&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build SAMTOOLS MERGE command</span>
    <span class="n">samtools_merge</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;merge&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-@&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">),</span>
        <span class="s2">&quot;-r&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_{alignment_name}_sorted_rg_merged.bam&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-b&quot;</span><span class="p">,</span> <span class="n">bam_list</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">samtools_merge</span>

    
<span class="k">def</span> <span class="nf">rename_to_merged_bam</span><span class="p">(</span><span class="n">rg_bam</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">alignment_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Renames the given RG BAM file to a merged and sorted BAM file name as if it was the result of merging and resorting of a bamlist.</span>

<span class="sd">    Args:</span>
<span class="sd">    - rg_bam (str): The path and name of the input RG BAM file.</span>
<span class="sd">    - sample (str): The name of the sample.</span>
<span class="sd">    - alignment_name (str): The name of the alignment.</span>
<span class="sd">    - output_folder (str): The path of the output folder.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: The command to rename the file.</span>

<span class="sd">    Examples:</span>
<span class="sd">    &gt;&gt;&gt; rename_to_merged_bam(&quot;/path/to/sample1_alignment1_sorted_rg.bam&quot;, &quot;sample1&quot;, &quot;alignment1&quot;, &quot;/path/to/output_folder&quot;)</span>
<span class="sd">    [&quot;mv&quot;, &quot;/path/to/sample1_alignment1_sorted_rg.bam&quot;, &quot;/path/to/output_folder/sample1_alignment1_merged_sorted.bam&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build mv command to rename the file</span>
    <span class="n">mv_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;mv&quot;</span><span class="p">,</span>
        <span class="n">rg_bam</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_{alignment_name}_sorted_rg_merged_sorted.bam&quot;</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">mv_command</span>


<span class="k">def</span> <span class="nf">mark_dups</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_picard</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Marks duplicates in the input BAM file using Picard MarkDuplicates.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): The path and name of the input BAM file.</span>
<span class="sd">    - sample (str): The name of the sample.</span>
<span class="sd">    - call_picard (str): The command to call Picard.</span>
<span class="sd">    - output_folder (str): The path of the output folder.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: The command to run Picard MarkDuplicates.</span>

<span class="sd">    Raises:</span>
<span class="sd">    - TypeError: If the BAM file does not end with &#39;.bam&#39;.</span>

<span class="sd">    Examples:</span>
<span class="sd">        </span>
<span class="sd">    &gt;&gt;&gt; mark_dups(&quot;/path/to/sample1.bam&quot;, &quot;sample1&quot;, &quot;/path/to/picard.jar&quot;, &quot;/path/to/output_folder&quot;)</span>
<span class="sd">    [&quot;/path/to/picard.jar&quot;, &quot;MarkDuplicates&quot;, &quot;METRICS_FILE=/path/to/output_folder/sample1_rmdup.txt&quot;, &quot;I=/path/to/sample1.bam&quot;, &quot;O=/path/to/output_folder/sample1_rmdup.bam&quot;, &quot;MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=800&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam!&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_rmdup.bam&#39;</span>

    <span class="c1"># Build Picard MarkDuplicates command</span>
    <span class="n">picard_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_picard</span><span class="p">,</span> <span class="s2">&quot;MarkDuplicates&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;METRICS_FILE={output_folder}/{sample}_rmdup.txt&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;I={bam_file}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;O={out_bam}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=800&quot;</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">picard_command</span>


<span class="k">def</span> <span class="nf">index_bam</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_samtools</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Indexes the input BAM file using SAMtools index.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): The path and name of the input BAM file.</span>
<span class="sd">    - call_samtools (str): The command to call SAMtools.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: The command to run SAMtools index.</span>

<span class="sd">    Examples:</span>
<span class="sd">    &gt;&gt;&gt; index_bam(&quot;/path/to/sample1.bam&quot;, &quot;samtools&quot;)</span>
<span class="sd">    [&quot;samtools&quot;, &quot;index&quot;, &quot;/path/to/sample1.bam&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build SAMtools index command</span>
    <span class="n">samtools_index</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;index&quot;</span><span class="p">,</span>
        <span class="n">bam_file</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">samtools_index</span>


<span class="k">def</span> <span class="nf">realigner_target_creator</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_gatk</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a GATK RealignerTargetCreator command to generate a list of intervals to be used for local realignment of</span>
<span class="sd">    reads.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): Path to input BAM file</span>
<span class="sd">    - sample (str): Sample name</span>
<span class="sd">    - call_gatk (str): Path to GATK executable</span>
<span class="sd">    - reference_fasta (str): Path to reference FASTA file</span>
<span class="sd">    - output_folder (str): Path to output folder</span>
<span class="sd">    - threads (int): Number of threads to use (default=1)</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list containing the GATK RealignerTargetCreator command and its arguments.</span>

<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; bam_file = &quot;/path/to/input.bam&quot;</span>
<span class="sd">    &gt;&gt;&gt; sample = &quot;my_sample&quot;</span>
<span class="sd">    &gt;&gt;&gt; output_folder = &quot;/path/to/output/folder&quot;</span>
<span class="sd">    &gt;&gt;&gt; threads = 4</span>
<span class="sd">    &gt;&gt;&gt; reference_fasta = &quot;/path/to/reference.fasta&quot;</span>
<span class="sd">    &gt;&gt;&gt; gatk_path = &quot;/path/to/gatk.jar&quot;</span>
<span class="sd">    &gt;&gt;&gt; gatk_command = realigner_target_creator(bam_file, sample, output_folder, threads, reference_fasta, gatk_path)</span>
<span class="sd">    &gt;&gt;&gt; print(gatk_command)</span>
<span class="sd">    [&#39;java&#39;, &#39;-jar&#39;, &#39;/path/to/gatk.jar&#39;, &#39;-T&#39;, &#39;RealignerTargetCreator&#39;, &#39;-nt&#39;, &#39;4&#39;, &#39;-R&#39;, &#39;/path/to/reference.fasta&#39;, &#39;-I&#39;, &#39;/path/to/input.bam&#39;, &#39;-o&#39;, &#39;/path/to/output/folder/my_sample_realignertargetcreator.intervals&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># build GATK RealignerTargetCreator command</span>
    <span class="n">gatk_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_gatk</span><span class="p">,</span> <span class="s2">&quot;-T&quot;</span><span class="p">,</span> <span class="s2">&quot;RealignerTargetCreator&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-nt&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">),</span>
        <span class="s2">&quot;-R&quot;</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">,</span>
        <span class="s2">&quot;-I&quot;</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_realignertargetcreator.intervals&quot;</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">gatk_command</span>


<span class="k">def</span> <span class="nf">indel_realigner</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">intervals</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_gatk</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Realigns indels in the specified BAM file using GATK&#39;s IndelRealigner.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): The BAM file to realign.</span>
<span class="sd">    - intervals (str): The intervals file produced by GATK&#39;s RealignerTargetCreator.</span>
<span class="sd">    - call_gatk (str): The path to the GATK executable.</span>
<span class="sd">    - reference_fasta (str): The path to the reference FASTA file.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: A list representing the GATK IndelRealigner command.</span>

<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; bam_file = &quot;/path/to/input.bam&quot;</span>
<span class="sd">    &gt;&gt;&gt; intervals_file = &quot;/path/to/intervals.intervals&quot;</span>
<span class="sd">    &gt;&gt;&gt; call_gatk = &quot;/path/to/gatk&quot;</span>
<span class="sd">    &gt;&gt;&gt; reference_fasta = &quot;/path/to/reference.fasta&quot;</span>
<span class="sd">    &gt;&gt;&gt; indel_realigner(bam_file, intervals_file, call_gatk, reference_fasta)</span>
<span class="sd">    [&quot;/path/to/gatk&quot;, &quot;-T&quot;, &quot;IndelRealigner&quot;, &quot;-R&quot;, &quot;/path/to/reference.fasta&quot;, </span>
<span class="sd">    &quot;-targetIntervals&quot;, &quot;/path/to/intervals.intervals&quot;, &quot;-I&quot;, &quot;/path/to/input.bam&quot;, </span>
<span class="sd">    &quot;-o&quot;, &quot;/path/to/input_indelrealigner.bam&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam!&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_indelrealigner.bam&#39;</span>

    <span class="c1"># Build GATK IndelRealigner command</span>
    <span class="n">gatk_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_gatk</span><span class="p">,</span> <span class="s2">&quot;-T&quot;</span><span class="p">,</span> <span class="s2">&quot;IndelRealigner&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-R&quot;</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">,</span>
        <span class="s2">&quot;-targetIntervals&quot;</span><span class="p">,</span> <span class="n">intervals</span><span class="p">,</span>
        <span class="s2">&quot;-I&quot;</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">out_bam</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">gatk_command</span>


<span class="k">def</span> <span class="nf">print_sample_pipeline</span><span class="p">(</span><span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">run</span><span class="p">:</span> <span class="nb">object</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print a pipeline of commands to align, sort, add read groups, merge bams, mark duplicates,</span>
<span class="sd">    create realignment targets, and realign indels for a given sample.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - sample (str): Name of the sample to be processed.</span>
<span class="sd">    - run (object): Object containing various pipeline parameters.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - None</span>
<span class="sd">    </span>
<span class="sd">    Example:</span>
<span class="sd">    print_sample_pipeline(sample=&quot;sample1&quot;, run=run)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;# Align reads of {sample}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">fastq_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">run</span><span class="o">.</span><span class="n">sample_dict</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    
    <span class="k">for</span> <span class="n">fastq_id</span> <span class="ow">in</span> <span class="n">fastq_ids</span><span class="p">:</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># align reads from files &#39;{run.sample_dict[sample][fastq_id][1]}&#39; and &#39;{run.sample_dict[sample][fastq_id][2]}&#39;:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">map_reads_command</span> <span class="o">=</span> <span class="n">map_reads</span><span class="p">(</span><span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">fastq_id</span> <span class="o">=</span> <span class="n">fastq_id</span><span class="p">,</span> <span class="n">run</span> <span class="o">=</span> <span class="n">run</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">map_reads_command</span><span class="p">)</span>
        <span class="n">first_bam</span> <span class="o">=</span> <span class="n">map_reads_command</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {first_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># sort the resulting bam file:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">sort_first_bam</span> <span class="o">=</span> <span class="n">sort_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">first_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sort_first_bam</span><span class="p">)</span>
        <span class="n">sorted_bam</span> <span class="o">=</span> <span class="n">sort_first_bam</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {sorted_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># add read groups to the sorted bam file:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">add_rgs_to_sorted</span> <span class="o">=</span> <span class="n">add_rg</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">sorted_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">fastq_id</span> <span class="o">=</span> <span class="n">fastq_id</span><span class="p">,</span> <span class="n">call_picard</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_picard</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">add_rgs_to_sorted</span><span class="p">)</span>
        <span class="n">rg_bam</span> <span class="o">=</span> <span class="n">add_rgs_to_sorted</span><span class="p">[</span><span class="mi">3</span><span class="p">][</span><span class="mi">2</span><span class="p">:]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {rg_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fastq_ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># {sample}&#39;s bams should be merged!</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># building {sample}&#39;s list of bams:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">make_bamlist</span> <span class="o">=</span> <span class="n">make_rg_bamlist</span><span class="p">(</span><span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">make_bamlist</span><span class="p">)</span>
        <span class="n">bam_list</span> <span class="o">=</span> <span class="n">make_bamlist</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {bam_list}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># merging {sample}&#39;s bams:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">merge_sample_bams</span> <span class="o">=</span> <span class="n">merge_bams</span><span class="p">(</span><span class="n">bam_list</span> <span class="o">=</span> <span class="n">bam_list</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span>
                                   <span class="n">alignment_name</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">alignment_name</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merge_sample_bams</span><span class="p">)</span>
        <span class="n">merged_bam</span> <span class="o">=</span> <span class="n">merge_sample_bams</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {merged_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># sorting {sample}&#39;s merged bam:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">sort_merged_bam</span> <span class="o">=</span> <span class="n">sort_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">merged_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sort_merged_bam</span><span class="p">)</span>
        <span class="n">merged_bam</span> <span class="o">=</span> <span class="n">sort_merged_bam</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {merged_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
    <span class="k">else</span><span class="p">:</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># {sample} only has one fastqid!</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;# renaming its only sorted_rg BAM to merged_sorted:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">rename</span> <span class="o">=</span> <span class="n">rename_to_merged_bam</span><span class="p">(</span><span class="n">rg_bam</span> <span class="o">=</span> <span class="n">rg_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">alignment_name</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">alignment_name</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rename</span><span class="p">)</span>
        <span class="n">merged_bam</span> <span class="o">=</span> <span class="n">rename</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {merged_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># marking duplicates of merged bam:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">md_command</span> <span class="o">=</span> <span class="n">mark_dups</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">merged_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">call_picard</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_picard</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">md_command</span><span class="p">)</span>
    <span class="n">md_bam</span> <span class="o">=</span> <span class="n">md_command</span><span class="p">[</span><span class="mi">4</span><span class="p">][</span><span class="mi">2</span><span class="p">:]</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {md_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># index the duplicate marked bam for gatk:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">index_cmd</span> <span class="o">=</span> <span class="n">index_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">md_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">index_cmd</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {md_bam}.bai</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># identify indel realignment targets:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">real_tar</span> <span class="o">=</span> <span class="n">realigner_target_creator</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">md_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">call_gatk</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_gatk</span><span class="p">,</span> <span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">real_tar</span><span class="p">)</span>
    <span class="n">target_file</span> <span class="o">=</span> <span class="n">real_tar</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {target_file}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># realign indels:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">ind_real</span> <span class="o">=</span> <span class="n">indel_realigner</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">md_bam</span><span class="p">,</span> <span class="n">intervals</span> <span class="o">=</span> <span class="n">target_file</span><span class="p">,</span> <span class="n">call_gatk</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_gatk</span><span class="p">,</span> <span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ind_real</span><span class="p">)</span>
    <span class="n">real_bam</span> <span class="o">=</span> <span class="n">ind_real</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {real_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># index the indel realigned bam for future analyses:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">index_cmd_2</span> <span class="o">=</span> <span class="n">index_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">real_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">index_cmd_2</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {md_bam}.bai</span><span class="se">\n</span><span class="s2">&quot;</span>
        
    <span class="k">return</span> <span class="n">bash_script</span>
</pre></div>

        </details>

            </section>
                <section id="RunConfig">
                                <div class="attr class">
        <a class="headerlink" href="#RunConfig">#&nbsp;&nbsp</a>

        
        <span class="def">class</span>
        <span class="name">RunConfig</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">class</span> <span class="nc">RunConfig</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A class to hold the configuration parameters for a run.</span>
<span class="sd">    </span>
<span class="sd">    Attributes:</span>
<span class="sd">    - sample_dict (dict): A dictionary mapping sample names to dictionaries mapping unique fastq_ids to the file names of the paired-end fastq files (r1 and r2).</span>
<span class="sd">    - samples (list): A list of sample names in the order they appear in the configuration file (within the dictionary sample_dict).</span>
<span class="sd">    - fastq_folder (str): The folder path where the fastq files are located.</span>
<span class="sd">    - reference_fasta (str): The path to the reference fasta file used for alignment.</span>
<span class="sd">    - output_folder (str): The folder path where the output files will be written.</span>
<span class="sd">    - threads (int): The number of threads to use for parallelized tools.</span>
<span class="sd">    - alignment_name (str): A label used to identify the alignment in the output file names.</span>
<span class="sd">    - call_bwa (str): The command or path to the executable for the BWA aligner.</span>
<span class="sd">    - call_samtools (str): The command or path to the executable for the SAMtools toolkit.</span>
<span class="sd">    - call_picard (str): The command or path to the executable for the Picard toolkit.</span>
<span class="sd">    - call_gatk (str): The command or path to the executable for the GATK toolkit.</span>
<span class="sd">    </span>
<span class="sd">    Methods:</span>
<span class="sd">    - __init__(self, config_file): Initializes a new RunConfig object from a YAML configuration file.</span>
<span class="sd">    - verify_arguments(self): Checks if the various elements of the configuration file exist</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">config_file</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes a new RunConfig object from a YAML configuration file.</span>
<span class="sd">        </span>
<span class="sd">        Parameters:</span>
<span class="sd">        - config_file (str): The path to the YAML configuration file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span> <span class="k">as</span> <span class="nb">file</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="n">Loader</span><span class="o">=</span><span class="n">FullLoader</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_dict</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;sample_dict&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;reference_fasta&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_folder</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;output_folder&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">threads</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;threads&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment_name</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;alignment_name&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_bwa</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_bwa&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_samtools</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_samtools&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_picard</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_picard&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_gatk</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_gatk&#39;</span><span class="p">]</span>        

    <span class="k">def</span> <span class="nf">verify_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Checks if the various elements of the configuration file exist</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># check if path to fastqs exists</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">)</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="s2">&quot;reference genome {self.reference_fasta} not found!&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;reference genome found!&quot;</span><span class="p">)</span>
</pre></div>

        </details>

            <div class="docstring"><p>A class to hold the configuration parameters for a run.</p>

<p>Attributes:</p>

<ul>
<li>sample_dict (dict): A dictionary mapping sample names to dictionaries mapping unique fastq_ids to the file names of the paired-end fastq files (r1 and r2).</li>
<li>samples (list): A list of sample names in the order they appear in the configuration file (within the dictionary sample_dict).</li>
<li>fastq_folder (str): The folder path where the fastq files are located.</li>
<li>reference_fasta (str): The path to the reference fasta file used for alignment.</li>
<li>output_folder (str): The folder path where the output files will be written.</li>
<li>threads (int): The number of threads to use for parallelized tools.</li>
<li>alignment_name (str): A label used to identify the alignment in the output file names.</li>
<li>call_bwa (str): The command or path to the executable for the BWA aligner.</li>
<li>call_samtools (str): The command or path to the executable for the SAMtools toolkit.</li>
<li>call_picard (str): The command or path to the executable for the Picard toolkit.</li>
<li>call_gatk (str): The command or path to the executable for the GATK toolkit.</li>
</ul>

<p>Methods:</p>

<ul>
<li>__init__(self, config_file): Initializes a new RunConfig object from a YAML configuration file.</li>
<li>verify_arguments(self): Checks if the various elements of the configuration file exist</li>
</ul>
</div>


                            <div id="RunConfig.__init__" class="classattr">
                                        <div class="attr function"><a class="headerlink" href="#RunConfig.__init__">#&nbsp;&nbsp</a>

        
            <span class="name">RunConfig</span><span class="signature">(config_file)</span>
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">config_file</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initializes a new RunConfig object from a YAML configuration file.</span>
<span class="sd">        </span>
<span class="sd">        Parameters:</span>
<span class="sd">        - config_file (str): The path to the YAML configuration file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">config_file</span><span class="p">)</span> <span class="k">as</span> <span class="nb">file</span><span class="p">:</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">yaml</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="n">Loader</span><span class="o">=</span><span class="n">FullLoader</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_dict</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;sample_dict&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="n">sample</span> <span class="k">for</span> <span class="n">sample</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;reference_fasta&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_folder</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;output_folder&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">threads</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;threads&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment_name</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;alignment_name&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_bwa</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_bwa&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_samtools</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_samtools&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_picard</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_picard&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">call_gatk</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;call_gatk&#39;</span><span class="p">]</span>        
</pre></div>

        </details>

            <div class="docstring"><p>Initializes a new RunConfig object from a YAML configuration file.</p>

<p>Parameters:</p>

<ul>
<li>config_file (str): The path to the YAML configuration file.</li>
</ul>
</div>


                            </div>
                            <div id="RunConfig.verify_arguments" class="classattr">
                                        <div class="attr function"><a class="headerlink" href="#RunConfig.verify_arguments">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">verify_arguments</span><span class="signature">(self)</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span>    <span class="k">def</span> <span class="nf">verify_arguments</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Checks if the various elements of the configuration file exist</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># check if path to fastqs exists</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">)</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="s2">&quot;reference genome {self.reference_fasta} not found!&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;reference genome found!&quot;</span><span class="p">)</span>
</pre></div>

        </details>

            <div class="docstring"><p>Checks if the various elements of the configuration file exist</p>
</div>


                            </div>
                </section>
                <section id="map_reads">
                            <div class="attr function"><a class="headerlink" href="#map_reads">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">map_reads</span><span class="signature">(fastq_id: str, sample: str, run: <a href="#RunConfig">src.RunConfig</a>) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">map_reads</span><span class="p">(</span><span class="n">fastq_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">run</span><span class="p">:</span> <span class="n">RunConfig</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Map reads from a pair of FASTQ files to a reference genome using BWA-MEM and SAMtools.</span>
<span class="sd">    </span>
<span class="sd">    Parameters:</span>
<span class="sd">    - fastq_id (str): The ID of the r1 and r2 FASTQ files pair to map reads from.</span>
<span class="sd">    - sample (str): The name of the sample to map reads for.</span>
<span class="sd">    - run (RunConfig): An object containing the parameters and file paths for the mapping run - created with the RunConfig() class.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list of command-line arguments for running BWA and SAMtools to map reads.</span>
<span class="sd">    </span>
<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; run = RunConfig(config/run.yml)</span>
<span class="sd">    &gt;&gt;&gt; map_reads(&#39;id1&#39;, &#39;sample1&#39;, run)</span>
<span class="sd">    [&#39;/path/to/bwa mem /path/to/reference.fasta /path/to/id1.fastq.gz /path/to/id1_2.fastq.gz -t 8 | /path/to/samtools view -hbS -@ 8 - -o /path/to/sample1_id1.bwa.bam&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># build BWA MEM + SAMTOOLS VIEW command</span>
    <span class="n">bwa_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">run</span><span class="o">.</span><span class="n">call_bwa</span><span class="p">,</span> <span class="s2">&quot;mem&quot;</span><span class="p">,</span> <span class="n">run</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{run.sample_dict[sample][fastq_id][0]}/{run.sample_dict[sample][fastq_id][1]}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{run.sample_dict[sample][fastq_id][0]}/{run.sample_dict[sample][fastq_id][2]}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">),</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;view&quot;</span><span class="p">,</span> <span class="s2">&quot;-hbS&quot;</span><span class="p">,</span> <span class="s2">&quot;-@&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">),</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{run.output_folder}/{sample}_{fastq_id}_{run.alignment_name}.bam&quot;</span>
    <span class="p">]</span>
    
    <span class="k">return</span> <span class="n">bwa_command</span>
</pre></div>

        </details>

            <div class="docstring"><p>Map reads from a pair of FASTQ files to a reference genome using BWA-MEM and SAMtools.</p>

<p>Parameters:</p>

<ul>
<li>fastq_id (str): The ID of the r1 and r2 FASTQ files pair to map reads from.</li>
<li>sample (str): The name of the sample to map reads for.</li>
<li>run (RunConfig): An object containing the parameters and file paths for the mapping run - created with the RunConfig() class.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: A list of command-line arguments for running BWA and SAMtools to map reads.</li>
</ul>

<p>Example:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">run</span> <span class="o">=</span> <span class="n">RunConfig</span><span class="p">(</span><span class="n">config</span><span class="o">/</span><span class="n">run</span><span class="o">.</span><span class="n">yml</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">map_reads</span><span class="p">(</span><span class="s1">&#39;id1&#39;</span><span class="p">,</span> <span class="s1">&#39;sample1&#39;</span><span class="p">,</span> <span class="n">run</span><span class="p">)</span>
<span class="go">[&#39;/path/to/bwa mem /path/to/reference.fasta /path/to/id1.fastq.gz /path/to/id1_2.fastq.gz -t 8 | /path/to/samtools view -hbS -@ 8 - -o /path/to/sample1_id1.bwa.bam&#39;]</span>
</code></pre></div>
</div>


                </section>
                <section id="sort_bam">
                            <div class="attr function"><a class="headerlink" href="#sort_bam">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">sort_bam</span><span class="signature">(bam_file: str, call_samtools: str, threads: int = 1) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">sort_bam</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_samtools</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Sort a BAM file using SAMtools sort command.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): Path to the BAM file to be sorted.</span>
<span class="sd">    - call_samtools (str): Path to the SAMtools executable.</span>
<span class="sd">    - threads (int): Number of threads to use for sorting the BAM file (default=1).</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list of command-line arguments for running the SAMtools sort command to sort the input BAM file.</span>
<span class="sd">    </span>
<span class="sd">    Raises:</span>
<span class="sd">    - TypeError: If the BAM file does not end with &#39;.bam&#39;.</span>

<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; sort_bam(&#39;/path/to/input.bam&#39;, &#39;/path/to/samtools&#39;, 4)</span>
<span class="sd">    [&#39;/path/to/samtools sort -@ 4 /path/to/input.bam -o /path/to/input_sorted.bam&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span> 
    
    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam !&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_sorted.bam&#39;</span>
    
    <span class="c1"># Build SAMTOOLS SORT command</span>
    <span class="n">samtools_sort</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;sort&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-@&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">),</span>
        <span class="n">bam_file</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">out_bam</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">samtools_sort</span>
</pre></div>

        </details>

            <div class="docstring"><p>Sort a BAM file using SAMtools sort command.</p>

<p>Args:</p>

<ul>
<li>bam_file (str): Path to the BAM file to be sorted.</li>
<li>call_samtools (str): Path to the SAMtools executable.</li>
<li>threads (int): Number of threads to use for sorting the BAM file (default=1).</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: A list of command-line arguments for running the SAMtools sort command to sort the input BAM file.</li>
</ul>

<p>Raises:</p>

<ul>
<li>TypeError: If the BAM file does not end with '.bam'.</li>
</ul>

<p>Example:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">sort_bam</span><span class="p">(</span><span class="s1">&#39;/path/to/input.bam&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/samtools&#39;</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="go">[&#39;/path/to/samtools sort -@ 4 /path/to/input.bam -o /path/to/input_sorted.bam&#39;]</span>
</code></pre></div>
</div>


                </section>
                <section id="add_rg">
                            <div class="attr function"><a class="headerlink" href="#add_rg">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">add_rg</span><span class="signature">(bam_file: str, sample: str, fastq_id: str, call_picard: str) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">add_rg</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">fastq_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_picard</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add read groups to a BAM file using PicardTools AddOrReplaceReadGroups.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): Path to the input BAM file to add read groups to.</span>
<span class="sd">    - sample (str): Name of the sample to add read groups for.</span>
<span class="sd">    - fastq_id (str): ID of the FASTQ file to add read groups for.</span>
<span class="sd">    - call_picard (str): Path to the PicardTools executable to use.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list of command-line arguments for running PicardTools AddOrReplaceReadGroups.</span>

<span class="sd">    Raises:</span>
<span class="sd">    - TypeError: If the BAM file does not end with &#39;.bam&#39;.</span>
<span class="sd">    </span>
<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; call_picard = &#39;/path/to/picard.jar&#39;</span>
<span class="sd">    &gt;&gt;&gt; bam_file = &#39;/path/to/input.bam&#39;</span>
<span class="sd">    &gt;&gt;&gt; sample = &#39;sample1&#39;</span>
<span class="sd">    &gt;&gt;&gt; fastq_id = &#39;fastq1&#39;</span>
<span class="sd">    &gt;&gt;&gt; add_rg(bam_file, sample, fastq_id, call_picard)</span>
<span class="sd">    [&#39;/path/to/picard.jar&#39;, &#39;AddOrReplaceReadGroups&#39;, &#39;I=/path/to/input.bam&#39;, &#39;O=/path/to/input_rg.bam&#39;, &#39;RGID=fastq1&#39;, &#39;RGLB=sample1_lib&#39;, &#39;RGPL=Illumina&#39;, &#39;RGPU=fastq1&#39;, &#39;RGSM=sample1&#39;, &#39;VALIDATION_STRINGENCY=SILENT&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam !&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_rg.bam&#39;</span>
    
    <span class="c1"># Build PICARDTOOLS AddOrReplaceReadGroups command</span>
    <span class="n">picard_add_rg</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_picard</span><span class="p">,</span> <span class="s2">&quot;AddOrReplaceReadGroups&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;I={bam_file}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;O={out_bam}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGID={fastq_id}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGLB={sample}_lib&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGPL=Illumina&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGPU={fastq_id}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;RGSM={sample}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;VALIDATION_STRINGENCY=SILENT&quot;</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">picard_add_rg</span>
</pre></div>

        </details>

            <div class="docstring"><p>Add read groups to a BAM file using PicardTools AddOrReplaceReadGroups.</p>

<p>Args:</p>

<ul>
<li>bam_file (str): Path to the input BAM file to add read groups to.</li>
<li>sample (str): Name of the sample to add read groups for.</li>
<li>fastq_id (str): ID of the FASTQ file to add read groups for.</li>
<li>call_picard (str): Path to the PicardTools executable to use.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: A list of command-line arguments for running PicardTools AddOrReplaceReadGroups.</li>
</ul>

<p>Raises:</p>

<ul>
<li>TypeError: If the BAM file does not end with '.bam'.</li>
</ul>

<p>Example:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">call_picard</span> <span class="o">=</span> <span class="s1">&#39;/path/to/picard.jar&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bam_file</span> <span class="o">=</span> <span class="s1">&#39;/path/to/input.bam&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span> <span class="o">=</span> <span class="s1">&#39;sample1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fastq_id</span> <span class="o">=</span> <span class="s1">&#39;fastq1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">add_rg</span><span class="p">(</span><span class="n">bam_file</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">fastq_id</span><span class="p">,</span> <span class="n">call_picard</span><span class="p">)</span>
<span class="go">[&#39;/path/to/picard.jar&#39;, &#39;AddOrReplaceReadGroups&#39;, &#39;I=/path/to/input.bam&#39;, &#39;O=/path/to/input_rg.bam&#39;, &#39;RGID=fastq1&#39;, &#39;RGLB=sample1_lib&#39;, &#39;RGPL=Illumina&#39;, &#39;RGPU=fastq1&#39;, &#39;RGSM=sample1&#39;, &#39;VALIDATION_STRINGENCY=SILENT&#39;]</span>
</code></pre></div>
</div>


                </section>
                <section id="make_rg_bamlist">
                            <div class="attr function"><a class="headerlink" href="#make_rg_bamlist">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">make_rg_bamlist</span><span class="signature">(sample: str, output_folder: str) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">make_rg_bamlist</span><span class="p">(</span><span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Constructs a command to generate a list of BAM files with read group (RG) information</span>
<span class="sd">    for a given sample and output folder.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - sample (str): The name of the sample to generate the BAM file list for.</span>
<span class="sd">    - output_folder (str): The folder where the BAM files are stored.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list containing the ls command to generate the BAM file list.</span>
<span class="sd">    </span>
<span class="sd">    Examples:</span>
<span class="sd">    &gt;&gt;&gt; make_rg_bamlist(&#39;sample1&#39;, &#39;/path/to/output/folder&#39;)</span>
<span class="sd">    [&#39;ls&#39;, &#39;/path/to/output/folder/sample1_*_sorted_rg.bam&#39;, &#39;&gt;&#39;, &#39;/path/to/output/folder/sample1.bam.list&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build ls command</span>
    <span class="n">ls_bamlist</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;ls&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_*_sorted_rg.bam&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}.bam.list&quot;</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">ls_bamlist</span>
</pre></div>

        </details>

            <div class="docstring"><p>Constructs a command to generate a list of BAM files with read group (RG) information
for a given sample and output folder.</p>

<p>Args:</p>

<ul>
<li>sample (str): The name of the sample to generate the BAM file list for.</li>
<li>output_folder (str): The folder where the BAM files are stored.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: A list containing the ls command to generate the BAM file list.</li>
</ul>

<p>Examples:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">make_rg_bamlist</span><span class="p">(</span><span class="s1">&#39;sample1&#39;</span><span class="p">,</span> <span class="s1">&#39;/path/to/output/folder&#39;</span><span class="p">)</span>
<span class="go">[&#39;ls&#39;, &#39;/path/to/output/folder/sample1_*_sorted_rg.bam&#39;, &#39;&gt;&#39;, &#39;/path/to/output/folder/sample1.bam.list&#39;]</span>
</code></pre></div>
</div>


                </section>
                <section id="merge_bams">
                            <div class="attr function"><a class="headerlink" href="#merge_bams">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">merge_bams</span><span class="signature">(
    bam_list: str,
    sample: str,
    call_samtools: str,
    alignment_name: str,
    output_folder: str,
    threads: int = 1
) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">merge_bams</span><span class="p">(</span><span class="n">bam_list</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_samtools</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">alignment_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Constructs a command to merge multiple BAM files into a single BAM file for a given sample.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - bam_list (str): The path to a file containing a list of BAM files to merge.</span>
<span class="sd">    - sample (str): The name of the sample to merge the BAM files for.</span>
<span class="sd">    - call_samtools (str): The command to call SAMTOOLS.</span>
<span class="sd">    - alignment_name (str): The name of the alignment</span>
<span class="sd">    - threads (int): The number of threads to use for the merging process (default=1).</span>
<span class="sd">    - output_folder (str): The folder where the merged BAM file will be stored.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list containing the SAMTOOLS MERGE command to merge the BAM files.</span>

<span class="sd">    Examples:</span>
<span class="sd">        </span>
<span class="sd">    &gt;&gt;&gt; merge_bams(&quot;/path/to/bam.list&quot;, &quot;sample1&quot;, &quot;samtools&quot;, 8, &quot;/path/to/output/folder&quot;)</span>
<span class="sd">    [&#39;samtools&#39;, &#39;merge&#39;, &#39;-@&#39;, &#39;8&#39;, &#39;-r&#39;, &#39;/path/to/output/folder/sample1_merged.bam&#39;, &#39;-b&#39;, &#39;/path/to/bam.list&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build SAMTOOLS MERGE command</span>
    <span class="n">samtools_merge</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;merge&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-@&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">),</span>
        <span class="s2">&quot;-r&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_{alignment_name}_sorted_rg_merged.bam&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-b&quot;</span><span class="p">,</span> <span class="n">bam_list</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">samtools_merge</span>
</pre></div>

        </details>

            <div class="docstring"><p>Constructs a command to merge multiple BAM files into a single BAM file for a given sample.</p>

<p>Args:</p>

<ul>
<li>bam_list (str): The path to a file containing a list of BAM files to merge.</li>
<li>sample (str): The name of the sample to merge the BAM files for.</li>
<li>call_samtools (str): The command to call SAMTOOLS.</li>
<li>alignment_name (str): The name of the alignment</li>
<li>threads (int): The number of threads to use for the merging process (default=1).</li>
<li>output_folder (str): The folder where the merged BAM file will be stored.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: A list containing the SAMTOOLS MERGE command to merge the BAM files.</li>
</ul>

<p>Examples:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">merge_bams</span><span class="p">(</span><span class="s2">&quot;/path/to/bam.list&quot;</span><span class="p">,</span> <span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="s2">&quot;samtools&quot;</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="s2">&quot;/path/to/output/folder&quot;</span><span class="p">)</span>
<span class="go">[&#39;samtools&#39;, &#39;merge&#39;, &#39;-@&#39;, &#39;8&#39;, &#39;-r&#39;, &#39;/path/to/output/folder/sample1_merged.bam&#39;, &#39;-b&#39;, &#39;/path/to/bam.list&#39;]</span>
</code></pre></div>
</div>


                </section>
                <section id="rename_to_merged_bam">
                            <div class="attr function"><a class="headerlink" href="#rename_to_merged_bam">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">rename_to_merged_bam</span><span class="signature">(
    rg_bam: str,
    sample: str,
    alignment_name: str,
    output_folder: str
) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">rename_to_merged_bam</span><span class="p">(</span><span class="n">rg_bam</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">alignment_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Renames the given RG BAM file to a merged and sorted BAM file name as if it was the result of merging and resorting of a bamlist.</span>

<span class="sd">    Args:</span>
<span class="sd">    - rg_bam (str): The path and name of the input RG BAM file.</span>
<span class="sd">    - sample (str): The name of the sample.</span>
<span class="sd">    - alignment_name (str): The name of the alignment.</span>
<span class="sd">    - output_folder (str): The path of the output folder.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: The command to rename the file.</span>

<span class="sd">    Examples:</span>
<span class="sd">    &gt;&gt;&gt; rename_to_merged_bam(&quot;/path/to/sample1_alignment1_sorted_rg.bam&quot;, &quot;sample1&quot;, &quot;alignment1&quot;, &quot;/path/to/output_folder&quot;)</span>
<span class="sd">    [&quot;mv&quot;, &quot;/path/to/sample1_alignment1_sorted_rg.bam&quot;, &quot;/path/to/output_folder/sample1_alignment1_merged_sorted.bam&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build mv command to rename the file</span>
    <span class="n">mv_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;mv&quot;</span><span class="p">,</span>
        <span class="n">rg_bam</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_{alignment_name}_sorted_rg_merged_sorted.bam&quot;</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">mv_command</span>
</pre></div>

        </details>

            <div class="docstring"><p>Renames the given RG BAM file to a merged and sorted BAM file name as if it was the result of merging and resorting of a bamlist.</p>

<p>Args:</p>

<ul>
<li>rg_bam (str): The path and name of the input RG BAM file.</li>
<li>sample (str): The name of the sample.</li>
<li>alignment_name (str): The name of the alignment.</li>
<li>output_folder (str): The path of the output folder.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: The command to rename the file.</li>
</ul>

<p>Examples:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">rename_to_merged_bam</span><span class="p">(</span><span class="s2">&quot;/path/to/sample1_alignment1_sorted_rg.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="s2">&quot;alignment1&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/output_folder&quot;</span><span class="p">)</span>
<span class="go">[&quot;mv&quot;, &quot;/path/to/sample1_alignment1_sorted_rg.bam&quot;, &quot;/path/to/output_folder/sample1_alignment1_merged_sorted.bam&quot;]</span>
</code></pre></div>
</div>


                </section>
                <section id="mark_dups">
                            <div class="attr function"><a class="headerlink" href="#mark_dups">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">mark_dups</span><span class="signature">(
    bam_file: str,
    sample: str,
    call_picard: str,
    output_folder: str
) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">mark_dups</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_picard</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Marks duplicates in the input BAM file using Picard MarkDuplicates.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): The path and name of the input BAM file.</span>
<span class="sd">    - sample (str): The name of the sample.</span>
<span class="sd">    - call_picard (str): The command to call Picard.</span>
<span class="sd">    - output_folder (str): The path of the output folder.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: The command to run Picard MarkDuplicates.</span>

<span class="sd">    Raises:</span>
<span class="sd">    - TypeError: If the BAM file does not end with &#39;.bam&#39;.</span>

<span class="sd">    Examples:</span>
<span class="sd">        </span>
<span class="sd">    &gt;&gt;&gt; mark_dups(&quot;/path/to/sample1.bam&quot;, &quot;sample1&quot;, &quot;/path/to/picard.jar&quot;, &quot;/path/to/output_folder&quot;)</span>
<span class="sd">    [&quot;/path/to/picard.jar&quot;, &quot;MarkDuplicates&quot;, &quot;METRICS_FILE=/path/to/output_folder/sample1_rmdup.txt&quot;, &quot;I=/path/to/sample1.bam&quot;, &quot;O=/path/to/output_folder/sample1_rmdup.bam&quot;, &quot;MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=800&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam!&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_rmdup.bam&#39;</span>

    <span class="c1"># Build Picard MarkDuplicates command</span>
    <span class="n">picard_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_picard</span><span class="p">,</span> <span class="s2">&quot;MarkDuplicates&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;METRICS_FILE={output_folder}/{sample}_rmdup.txt&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;I={bam_file}&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="s2">&quot;O={out_bam}&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=800&quot;</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">picard_command</span>
</pre></div>

        </details>

            <div class="docstring"><p>Marks duplicates in the input BAM file using Picard MarkDuplicates.</p>

<p>Args:</p>

<ul>
<li>bam_file (str): The path and name of the input BAM file.</li>
<li>sample (str): The name of the sample.</li>
<li>call_picard (str): The command to call Picard.</li>
<li>output_folder (str): The path of the output folder.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: The command to run Picard MarkDuplicates.</li>
</ul>

<p>Raises:</p>

<ul>
<li>TypeError: If the BAM file does not end with '.bam'.</li>
</ul>

<p>Examples:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">mark_dups</span><span class="p">(</span><span class="s2">&quot;/path/to/sample1.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;sample1&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/picard.jar&quot;</span><span class="p">,</span> <span class="s2">&quot;/path/to/output_folder&quot;</span><span class="p">)</span>
<span class="go">[&quot;/path/to/picard.jar&quot;, &quot;MarkDuplicates&quot;, &quot;METRICS_FILE=/path/to/output_folder/sample1_rmdup.txt&quot;, &quot;I=/path/to/sample1.bam&quot;, &quot;O=/path/to/output_folder/sample1_rmdup.bam&quot;, &quot;MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=800&quot;]</span>
</code></pre></div>
</div>


                </section>
                <section id="index_bam">
                            <div class="attr function"><a class="headerlink" href="#index_bam">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">index_bam</span><span class="signature">(bam_file: str, call_samtools: str) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">index_bam</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_samtools</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Indexes the input BAM file using SAMtools index.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): The path and name of the input BAM file.</span>
<span class="sd">    - call_samtools (str): The command to call SAMtools.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: The command to run SAMtools index.</span>

<span class="sd">    Examples:</span>
<span class="sd">    &gt;&gt;&gt; index_bam(&quot;/path/to/sample1.bam&quot;, &quot;samtools&quot;)</span>
<span class="sd">    [&quot;samtools&quot;, &quot;index&quot;, &quot;/path/to/sample1.bam&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Build SAMtools index command</span>
    <span class="n">samtools_index</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_samtools</span><span class="p">,</span> <span class="s2">&quot;index&quot;</span><span class="p">,</span>
        <span class="n">bam_file</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">samtools_index</span>
</pre></div>

        </details>

            <div class="docstring"><p>Indexes the input BAM file using SAMtools index.</p>

<p>Args:</p>

<ul>
<li>bam_file (str): The path and name of the input BAM file.</li>
<li>call_samtools (str): The command to call SAMtools.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: The command to run SAMtools index.</li>
</ul>

<p>Examples:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">index_bam</span><span class="p">(</span><span class="s2">&quot;/path/to/sample1.bam&quot;</span><span class="p">,</span> <span class="s2">&quot;samtools&quot;</span><span class="p">)</span>
<span class="go">[&quot;samtools&quot;, &quot;index&quot;, &quot;/path/to/sample1.bam&quot;]</span>
</code></pre></div>
</div>


                </section>
                <section id="realigner_target_creator">
                            <div class="attr function"><a class="headerlink" href="#realigner_target_creator">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">realigner_target_creator</span><span class="signature">(
    bam_file: str,
    sample: str,
    call_gatk: str,
    reference_fasta: str,
    output_folder: str,
    threads: int = 1
) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">realigner_target_creator</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_gatk</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a GATK RealignerTargetCreator command to generate a list of intervals to be used for local realignment of</span>
<span class="sd">    reads.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): Path to input BAM file</span>
<span class="sd">    - sample (str): Sample name</span>
<span class="sd">    - call_gatk (str): Path to GATK executable</span>
<span class="sd">    - reference_fasta (str): Path to reference FASTA file</span>
<span class="sd">    - output_folder (str): Path to output folder</span>
<span class="sd">    - threads (int): Number of threads to use (default=1)</span>
<span class="sd">        </span>
<span class="sd">    Returns:</span>
<span class="sd">    - list: A list containing the GATK RealignerTargetCreator command and its arguments.</span>

<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; bam_file = &quot;/path/to/input.bam&quot;</span>
<span class="sd">    &gt;&gt;&gt; sample = &quot;my_sample&quot;</span>
<span class="sd">    &gt;&gt;&gt; output_folder = &quot;/path/to/output/folder&quot;</span>
<span class="sd">    &gt;&gt;&gt; threads = 4</span>
<span class="sd">    &gt;&gt;&gt; reference_fasta = &quot;/path/to/reference.fasta&quot;</span>
<span class="sd">    &gt;&gt;&gt; gatk_path = &quot;/path/to/gatk.jar&quot;</span>
<span class="sd">    &gt;&gt;&gt; gatk_command = realigner_target_creator(bam_file, sample, output_folder, threads, reference_fasta, gatk_path)</span>
<span class="sd">    &gt;&gt;&gt; print(gatk_command)</span>
<span class="sd">    [&#39;java&#39;, &#39;-jar&#39;, &#39;/path/to/gatk.jar&#39;, &#39;-T&#39;, &#39;RealignerTargetCreator&#39;, &#39;-nt&#39;, &#39;4&#39;, &#39;-R&#39;, &#39;/path/to/reference.fasta&#39;, &#39;-I&#39;, &#39;/path/to/input.bam&#39;, &#39;-o&#39;, &#39;/path/to/output/folder/my_sample_realignertargetcreator.intervals&#39;]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># build GATK RealignerTargetCreator command</span>
    <span class="n">gatk_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_gatk</span><span class="p">,</span> <span class="s2">&quot;-T&quot;</span><span class="p">,</span> <span class="s2">&quot;RealignerTargetCreator&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-nt&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">threads</span><span class="p">),</span>
        <span class="s2">&quot;-R&quot;</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">,</span>
        <span class="s2">&quot;-I&quot;</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{output_folder}/{sample}_realignertargetcreator.intervals&quot;</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">gatk_command</span>
</pre></div>

        </details>

            <div class="docstring"><p>Creates a GATK RealignerTargetCreator command to generate a list of intervals to be used for local realignment of
reads.</p>

<p>Args:</p>

<ul>
<li>bam_file (str): Path to input BAM file</li>
<li>sample (str): Sample name</li>
<li>call_gatk (str): Path to GATK executable</li>
<li>reference_fasta (str): Path to reference FASTA file</li>
<li>output_folder (str): Path to output folder</li>
<li>threads (int): Number of threads to use (default=1)</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: A list containing the GATK RealignerTargetCreator command and its arguments.</li>
</ul>

<p>Example:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">bam_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span> <span class="o">=</span> <span class="s2">&quot;my_sample&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s2">&quot;/path/to/output/folder&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">threads</span> <span class="o">=</span> <span class="mi">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_fasta</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gatk_path</span> <span class="o">=</span> <span class="s2">&quot;/path/to/gatk.jar&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gatk_command</span> <span class="o">=</span> <span class="n">realigner_target_creator</span><span class="p">(</span><span class="n">bam_file</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">threads</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">,</span> <span class="n">gatk_path</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">gatk_command</span><span class="p">)</span>
<span class="go">[&#39;java&#39;, &#39;-jar&#39;, &#39;/path/to/gatk.jar&#39;, &#39;-T&#39;, &#39;RealignerTargetCreator&#39;, &#39;-nt&#39;, &#39;4&#39;, &#39;-R&#39;, &#39;/path/to/reference.fasta&#39;, &#39;-I&#39;, &#39;/path/to/input.bam&#39;, &#39;-o&#39;, &#39;/path/to/output/folder/my_sample_realignertargetcreator.intervals&#39;]</span>
</code></pre></div>
</div>


                </section>
                <section id="indel_realigner">
                            <div class="attr function"><a class="headerlink" href="#indel_realigner">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">indel_realigner</span><span class="signature">(
    bam_file: str,
    intervals: str,
    call_gatk: str,
    reference_fasta: str
) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">indel_realigner</span><span class="p">(</span><span class="n">bam_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">intervals</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">call_gatk</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Realigns indels in the specified BAM file using GATK&#39;s IndelRealigner.</span>

<span class="sd">    Args:</span>
<span class="sd">    - bam_file (str): The BAM file to realign.</span>
<span class="sd">    - intervals (str): The intervals file produced by GATK&#39;s RealignerTargetCreator.</span>
<span class="sd">    - call_gatk (str): The path to the GATK executable.</span>
<span class="sd">    - reference_fasta (str): The path to the reference FASTA file.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - list: A list representing the GATK IndelRealigner command.</span>

<span class="sd">    Example:</span>
<span class="sd">    &gt;&gt;&gt; bam_file = &quot;/path/to/input.bam&quot;</span>
<span class="sd">    &gt;&gt;&gt; intervals_file = &quot;/path/to/intervals.intervals&quot;</span>
<span class="sd">    &gt;&gt;&gt; call_gatk = &quot;/path/to/gatk&quot;</span>
<span class="sd">    &gt;&gt;&gt; reference_fasta = &quot;/path/to/reference.fasta&quot;</span>
<span class="sd">    &gt;&gt;&gt; indel_realigner(bam_file, intervals_file, call_gatk, reference_fasta)</span>
<span class="sd">    [&quot;/path/to/gatk&quot;, &quot;-T&quot;, &quot;IndelRealigner&quot;, &quot;-R&quot;, &quot;/path/to/reference.fasta&quot;, </span>
<span class="sd">    &quot;-targetIntervals&quot;, &quot;/path/to/intervals.intervals&quot;, &quot;-I&quot;, &quot;/path/to/input.bam&quot;, </span>
<span class="sd">    &quot;-o&quot;, &quot;/path/to/input_indelrealigner.bam&quot;]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Check that bam_file ends with &#39;.bam&#39;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">bam_file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.bam&#39;</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;bam file should end in .bam!&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">out_bam</span> <span class="o">=</span> <span class="n">bam_file</span><span class="p">[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_indelrealigner.bam&#39;</span>

    <span class="c1"># Build GATK IndelRealigner command</span>
    <span class="n">gatk_command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">call_gatk</span><span class="p">,</span> <span class="s2">&quot;-T&quot;</span><span class="p">,</span> <span class="s2">&quot;IndelRealigner&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-R&quot;</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">,</span>
        <span class="s2">&quot;-targetIntervals&quot;</span><span class="p">,</span> <span class="n">intervals</span><span class="p">,</span>
        <span class="s2">&quot;-I&quot;</span><span class="p">,</span> <span class="n">bam_file</span><span class="p">,</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="n">out_bam</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">gatk_command</span>
</pre></div>

        </details>

            <div class="docstring"><p>Realigns indels in the specified BAM file using GATK's IndelRealigner.</p>

<p>Args:</p>

<ul>
<li>bam_file (str): The BAM file to realign.</li>
<li>intervals (str): The intervals file produced by GATK's RealignerTargetCreator.</li>
<li>call_gatk (str): The path to the GATK executable.</li>
<li>reference_fasta (str): The path to the reference FASTA file.</li>
</ul>

<p>Returns:</p>

<ul>
<li>list: A list representing the GATK IndelRealigner command.</li>
</ul>

<p>Example:</p>

<div class="pdoc-code codehilite"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">bam_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/input.bam&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">intervals_file</span> <span class="o">=</span> <span class="s2">&quot;/path/to/intervals.intervals&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">call_gatk</span> <span class="o">=</span> <span class="s2">&quot;/path/to/gatk&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">reference_fasta</span> <span class="o">=</span> <span class="s2">&quot;/path/to/reference.fasta&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">indel_realigner</span><span class="p">(</span><span class="n">bam_file</span><span class="p">,</span> <span class="n">intervals_file</span><span class="p">,</span> <span class="n">call_gatk</span><span class="p">,</span> <span class="n">reference_fasta</span><span class="p">)</span>
<span class="go">[&quot;/path/to/gatk&quot;, &quot;-T&quot;, &quot;IndelRealigner&quot;, &quot;-R&quot;, &quot;/path/to/reference.fasta&quot;, </span>
<span class="go">&quot;-targetIntervals&quot;, &quot;/path/to/intervals.intervals&quot;, &quot;-I&quot;, &quot;/path/to/input.bam&quot;, </span>
<span class="go">&quot;-o&quot;, &quot;/path/to/input_indelrealigner.bam&quot;]</span>
</code></pre></div>
</div>


                </section>
                <section id="print_sample_pipeline">
                            <div class="attr function"><a class="headerlink" href="#print_sample_pipeline">#&nbsp;&nbsp</a>

        
            <span class="def">def</span>
            <span class="name">print_sample_pipeline</span><span class="signature">(sample: str, run: object) -&gt; list</span>:
    </div>

            <details>
            <summary>View Source</summary>
            <div class="pdoc-code codehilite"><pre><span></span><span class="k">def</span> <span class="nf">print_sample_pipeline</span><span class="p">(</span><span class="n">sample</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">run</span><span class="p">:</span> <span class="nb">object</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Print a pipeline of commands to align, sort, add read groups, merge bams, mark duplicates,</span>
<span class="sd">    create realignment targets, and realign indels for a given sample.</span>
<span class="sd">    </span>
<span class="sd">    Args:</span>
<span class="sd">    - sample (str): Name of the sample to be processed.</span>
<span class="sd">    - run (object): Object containing various pipeline parameters.</span>
<span class="sd">    </span>
<span class="sd">    Returns:</span>
<span class="sd">    - None</span>
<span class="sd">    </span>
<span class="sd">    Example:</span>
<span class="sd">    print_sample_pipeline(sample=&quot;sample1&quot;, run=run)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;# Align reads of {sample}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">fastq_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">run</span><span class="o">.</span><span class="n">sample_dict</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
    
    <span class="k">for</span> <span class="n">fastq_id</span> <span class="ow">in</span> <span class="n">fastq_ids</span><span class="p">:</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># align reads from files &#39;{run.sample_dict[sample][fastq_id][1]}&#39; and &#39;{run.sample_dict[sample][fastq_id][2]}&#39;:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">map_reads_command</span> <span class="o">=</span> <span class="n">map_reads</span><span class="p">(</span><span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">fastq_id</span> <span class="o">=</span> <span class="n">fastq_id</span><span class="p">,</span> <span class="n">run</span> <span class="o">=</span> <span class="n">run</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">map_reads_command</span><span class="p">)</span>
        <span class="n">first_bam</span> <span class="o">=</span> <span class="n">map_reads_command</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {first_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># sort the resulting bam file:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">sort_first_bam</span> <span class="o">=</span> <span class="n">sort_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">first_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sort_first_bam</span><span class="p">)</span>
        <span class="n">sorted_bam</span> <span class="o">=</span> <span class="n">sort_first_bam</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {sorted_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>

        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># add read groups to the sorted bam file:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">add_rgs_to_sorted</span> <span class="o">=</span> <span class="n">add_rg</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">sorted_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">fastq_id</span> <span class="o">=</span> <span class="n">fastq_id</span><span class="p">,</span> <span class="n">call_picard</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_picard</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">add_rgs_to_sorted</span><span class="p">)</span>
        <span class="n">rg_bam</span> <span class="o">=</span> <span class="n">add_rgs_to_sorted</span><span class="p">[</span><span class="mi">3</span><span class="p">][</span><span class="mi">2</span><span class="p">:]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {rg_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fastq_ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># {sample}&#39;s bams should be merged!</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># building {sample}&#39;s list of bams:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">make_bamlist</span> <span class="o">=</span> <span class="n">make_rg_bamlist</span><span class="p">(</span><span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">make_bamlist</span><span class="p">)</span>
        <span class="n">bam_list</span> <span class="o">=</span> <span class="n">make_bamlist</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {bam_list}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># merging {sample}&#39;s bams:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">merge_sample_bams</span> <span class="o">=</span> <span class="n">merge_bams</span><span class="p">(</span><span class="n">bam_list</span> <span class="o">=</span> <span class="n">bam_list</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span>
                                   <span class="n">alignment_name</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">alignment_name</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">merge_sample_bams</span><span class="p">)</span>
        <span class="n">merged_bam</span> <span class="o">=</span> <span class="n">merge_sample_bams</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {merged_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># sorting {sample}&#39;s merged bam:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">sort_merged_bam</span> <span class="o">=</span> <span class="n">sort_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">merged_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sort_merged_bam</span><span class="p">)</span>
        <span class="n">merged_bam</span> <span class="o">=</span> <span class="n">sort_merged_bam</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {merged_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
        
    <span class="k">else</span><span class="p">:</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># {sample} only has one fastqid!</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;# renaming its only sorted_rg BAM to merged_sorted:</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="n">rename</span> <span class="o">=</span> <span class="n">rename_to_merged_bam</span><span class="p">(</span><span class="n">rg_bam</span> <span class="o">=</span> <span class="n">rg_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">alignment_name</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">alignment_name</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">rename</span><span class="p">)</span>
        <span class="n">merged_bam</span> <span class="o">=</span> <span class="n">rename</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {merged_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># marking duplicates of merged bam:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">md_command</span> <span class="o">=</span> <span class="n">mark_dups</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">merged_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">call_picard</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_picard</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">md_command</span><span class="p">)</span>
    <span class="n">md_bam</span> <span class="o">=</span> <span class="n">md_command</span><span class="p">[</span><span class="mi">4</span><span class="p">][</span><span class="mi">2</span><span class="p">:]</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {md_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># index the duplicate marked bam for gatk:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">index_cmd</span> <span class="o">=</span> <span class="n">index_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">md_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">index_cmd</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {md_bam}.bai</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># identify indel realignment targets:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">real_tar</span> <span class="o">=</span> <span class="n">realigner_target_creator</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">md_bam</span><span class="p">,</span> <span class="n">sample</span> <span class="o">=</span> <span class="n">sample</span><span class="p">,</span> <span class="n">call_gatk</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_gatk</span><span class="p">,</span> <span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">,</span> <span class="n">output_folder</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">output_folder</span><span class="p">,</span> <span class="n">threads</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">threads</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">real_tar</span><span class="p">)</span>
    <span class="n">target_file</span> <span class="o">=</span> <span class="n">real_tar</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {target_file}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># realign indels:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">ind_real</span> <span class="o">=</span> <span class="n">indel_realigner</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">md_bam</span><span class="p">,</span> <span class="n">intervals</span> <span class="o">=</span> <span class="n">target_file</span><span class="p">,</span> <span class="n">call_gatk</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_gatk</span><span class="p">,</span> <span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ind_real</span><span class="p">)</span>
    <span class="n">real_bam</span> <span class="o">=</span> <span class="n">ind_real</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {real_bam}</span><span class="se">\n</span><span class="s2">&quot;</span>
    
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># index the indel realigned bam for future analyses:</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="n">index_cmd_2</span> <span class="o">=</span> <span class="n">index_bam</span><span class="p">(</span><span class="n">bam_file</span> <span class="o">=</span> <span class="n">real_bam</span><span class="p">,</span> <span class="n">call_samtools</span> <span class="o">=</span> <span class="n">run</span><span class="o">.</span><span class="n">call_samtools</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">index_cmd_2</span><span class="p">)</span>
    <span class="n">bash_script</span> <span class="o">=</span> <span class="n">bash_script</span> <span class="o">+</span> <span class="n">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2"># -&gt; {md_bam}.bai</span><span class="se">\n</span><span class="s2">&quot;</span>
        
    <span class="k">return</span> <span class="n">bash_script</span>
</pre></div>

        </details>

            <div class="docstring"><p>Print a pipeline of commands to align, sort, add read groups, merge bams, mark duplicates,
create realignment targets, and realign indels for a given sample.</p>

<p>Args:</p>

<ul>
<li>sample (str): Name of the sample to be processed.</li>
<li>run (object): Object containing various pipeline parameters.</li>
</ul>

<p>Returns:</p>

<ul>
<li>None</li>
</ul>

<p>Example:
print_sample_pipeline(sample="sample1", run=run)</p>
</div>


                </section>
    </main>
</body>
</html>